### ISSUE TITLE
vo=caca playback breaks TTY

### ISSUE THREAD

**Issue by rr-** [mpv-player/mpv#1787]
1. Play anything with `vo=caca`
2. Quit, or finish playback normally
3. TTY is messed up, no input is shown, etc. Just like forcefully killing vim from inside vim.

Here's example SSH session:

![example](https://cloud.githubusercontent.com/assets/1045476/7072020/6e0a061e-dee9-11e4-86aa-96cda557283f.gif)

Confirmed the issue also with native Linux console. Inside X, playing through libcaca is most probably going to spawn new console window, so all the side effects will go unnoticed.

List of commands used in the GIF above, in order:

1.  `./build/mpv -vo=caca ./kill-la-kill-01.mkv --quiet`
2. `stty`
3. `ls`
4. `stty sane` (restores normal settings)
5. `ls`

**Comment by wm4** [comment]
It looks like we correctly terminate caca, though.

Anyway, the question is, should we care? This VO is at best a joke.

**Comment by rr-** [comment]
`aa` is already removed. IMO ASCII playback is nice to have and I wouldn't consider it bloatware (it's just one optional library and 300 sloc).

**Comment by rr-** [comment]
Confirmed `mplayer` to work fine. Quick glance at its code reveals that there are almost no differences, except mplayer handles subtitles and OSD gracefully.

**Comment by wm4** [comment]
>mplayer tries to render also subtitles and OSD, 

Yes, and this weird text support didn't seem worth it just for vo_caca after all.

>and mpv draws everything to intermediate buffer instead of passing it to libcaca directly.

Because libcaca doesn't accept an image stride.

**Comment by rr-** [comment]
Confirmed it's not libcaca's problem - compiled sample programs and they work fine. The call that triggers this behavior in mpv is `caca_create_display`, which works fine in a stand-alone program. I suspect mpv is doing something to terminal, like outputting some stuff. But even if I do `--really-quiet`, it doesn't change anything.
Will investigate further once I get home.

**Comment by wm4** [comment]
Try --no-terminal.

**Comment by wm4** [comment]
PS: and yes, it's doing something to the terminal - mostly status line output.

**Comment by rr-** [comment]
Nope - doesn't seem to help.

**Comment by wm4** [comment]
That's strange. I'm pretty sure mpv doesn't even touch the terminal (or any of the std streams) if this option is passed.

**Comment by rr-** [comment]
Found the culrpit: it's the `--input-terminal` option. By default it's on, turning it off with `--input-terminal=off` no longer breaks the TTY. You don't even need the other options (`--terminal` and `--quiet`).

Do we do anything about it, or mark the issue as closed?

**Comment by wm4** [comment]
--input-terminal should do nothing if terminal is disabled with --no-terminal. If it does, it's a bug.

**Comment by wm4** [comment]
>Do we do anything about it, or mark the issue as closed?

Up to you. I certainly don't care about caca.

**Comment by wm4** [comment]
>--input-terminal should do nothing if terminal is disabled with --no-terminal. If it does, it's a bug.

It was. I just fixed it.

**Comment by rr-** [comment]
`--no-terminal` controls `options.use_terminal` while `--no-input-terminal` controls `options.consolecontrols`.

The way I see, the fix could look like this...

    diff --git a/player/main.c b/player/main.c
    index 77ec805..a3236ed 100644
    --- a/player/main.c
    +++ b/player/main.c
    @@ -428,6 +428,10 @@ int mp_initialize(struct MPContext *mpctx, char **options)
             m_config_set_option0(mpctx->mconfig, "input-file", "/dev/stdin");
         }
    
    +    if (!opts->use_terminal) {
    +        opts->consolecontrols = 0;
    +    }
    +
         if (!mpctx->playlist->first && !opts->player_idle_mode)
             return -3;

...and I'm not sure how can I set some options from within a `vo`. It would need to happen before `mp_initialize` does the terminal stuff. Doing it in `main.c` is a clear breach of separation of concerns contract. Perhaps it's better to give this idea up altogether and require the user to type these settings manually so that no magic happens behind the scenes.

**Comment by wm4** [comment]
>I'm not sure how can I set some options from within a vo.

Not at all. It'd require some mechanism to deinit the terminal when the VO is created or so. (There's a similar mechanism with mp_msg_mute().)

> It would need to happen before mp_initialize does the terminal stuff. Doing it in main.c is a clear breach of separation of concerns contract. Perhaps it's better to give this idea up altogether and require the user to type these settings manually so that no magic happens behind the scenes.

It would be justified if it was a more important use-case, but yeah, just for caca I don't think it's worth to bother. The user can just use the ``reset`` command to restore the terminal anyway. Or maybe the mpv terminal uninit code could be more clever about it, I don't know.
